// @quarkly/atomize/build/cjs/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
var _exportNames = {
  dict: true,
  themeDefault: true,
  splitCSSRule: true,
  aliasesDict: true,
  getTransformer: true,
  transformVar: true
};
exports["default"] = exports.transformVar = exports.getTransformer = exports.aliasesDict = exports.splitCSSRule = exports.themeDefault = exports.dict = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _atomicStyles = require("./atomic-styles");

var _dict = _interopRequireWildcard(require("./constants/dict"));

exports.dict = _dict["default"];
exports.aliasesDict = _dict.hashPropsWithAliases;

var _themeDefault = _interopRequireDefault(require("./constants/themeDefault"));

exports.themeDefault = _themeDefault["default"];

var _splitCssRule = _interopRequireDefault(require("./atomic-styles/split-css-rule"));

exports.splitCSSRule = _splitCssRule["default"];

var _bootstrap = require("./atomic-styles/bootstrap");

exports.getTransformer = _bootstrap.getTransformer;
exports.transformVar = _bootstrap.transformVar;

var _propInfo = require("./prop-info");

Object.keys(_propInfo).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _propInfo[key];
});

var _default = (0, _atomicStyles.wrap)(_styledComponents["default"]);

exports["default"] = _default;
